src/cgr_gwas_qc/workflow/scripts/vcf2abf.py

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[1]:
+
+
+from pathlib import Path
+from typing import Optional
+
+import numpy as np
+import pandas as pd
+import typer
+from pysam import VariantFile
+
+app = typer.Typer(add_completion=False)
+
+
+# In[2]:
+
+
+def is_biallelic_snp(rec):
+    return len(rec.ref) == 1 and len(rec.alts[0]) == 1
+
+
+# In[27]:
+
+
+@app.command()
+def main(
+    bcf_file: Path = typer.Argument(
+        ..., help="Path to the sample VCF.", exists=True, readable=True
+    ),
+    kgvcf_file: Path = typer.Argument(
+        ..., help="Path to the 1000 genomes VCF file to get ABF.", exists=True, readable=True
+    ),
+    population: str = typer.Argument(
+        ...,
+        help="Which population in VCF file to use for ABF [Not implemented,global/AF used currently].",
+    ),
+    abf_file: Optional[Path] = typer.Argument(
+        None, help="Output file name. If none given it will use the `VCF_FILE + .abf.txt`."
+    ),
+):
+    """Pull out the B allele frequency for a given subpopulation.
+
+    Takes variants from the provided `vcf_file` and tries to find their B
+    allele frequency in the provided `kgvcf_file`. The B allele frequencies are
+    based on the allele frequency for the provided subpopulation
+    `population`. If the provided subpopulation doesn't exist, it will output
+    a list of available populations. If the variant does not exists in the
+    `vcf_file` then the ABF will be set to "NA".
+    """
+
+    if abf_file is None:
+        abf_file = bcf_file.with_suffix(".abf.txt")
+
+    # vcf_file='/scratch/rajwanir2/data_catalog/QC-test/results/GSAMD-24v1-0-contamseries/vcf/GSAMD-24v1-0-contamseries.bcf'
+    # abf_file='/scratch/rajwanir2/data_catalog/QC-test/GSAMD-24v1-0.new.abf'
+    # kg_vcf='/DCEG/CGF/Bioinformatics/Production/data/thousG/hg38/ALL.wgs.shapeit2_integrated_v1a.GRCh38.20181129.sites.vcf.gz'
+
+    # retrieving needed info from the input vcf
+    vcf = VariantFile(bcf_file, drop_samples=True)
+    vcf_info = []
+    for rec in vcf:
+        vcf_info.append(
+            [
+                rec.id,
+                rec.chrom,
+                rec.pos,
+                rec.ref[0],
+                rec.alts[0],
+                is_biallelic_snp(rec),
+                rec.info["GC_SCORE"],
```

Comment:
Does this line need to be edited too? I got an error again about the `GC_SCORE` field.
```
rule pull_b_allele_freq_from_1kg:
    input: /mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/samples.bcf, /DCEG/Projects/DataDelivery/GWAS_Data_Delivery/CONFLUENCE_FB28648_05172021/testing_500samples/manifest/ALL.wgs.shapeit2_integrated_v1a.GRCh38.20181129.sites.vcf.gz
    output: sample_level/GSAv3Confluence.AF.abf.txt
    jobid: 0                                                                                                                      resources: mem_mb=19787, disk_mb=19787, tmpdir=/tmp                                                                       
/home/marksja/miniconda3/envs/cgr-dev/bin/python /mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmph3hp_w1c.vcf2abf.py
[E::idx_find_and_load] Could not retrieve index file for '/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/samples.bcf'
Traceback (most recent call last):
  File "/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmph3hp_w1c.vcf2abf.py", line 144, in <module>
    main(**defaults)  # type: ignore
  File "/mnt/nfs/gigantor/ifs/DCEG/Home/marksja/dev-tests/024/.snakemake/scripts/tmph3hp_w1c.vcf2abf.py", line 77, in main
    rec.info["GC_SCORE"],
  File "pysam/libcbcf.pyx", line 2598, in pysam.libcbcf.VariantRecordInfo.__getitem__
KeyError: 'Unknown INFO field: GC_SCORE'
[Fri Sep 13 15:08:08 2024]                                                                                                    Error in rule pull_b_allele_freq_from_1kg:
    jobid: 0                                                                                                                      output: sample_level/GSAv3Confluence.AF.abf.txt

```

Reply:
Yes, for compatibility with gtc2vcf prepared bcf. 1ee63adf5d118aae96def3b0dba6c37614df3488 should fix this. Thanks.

Reply:
@jaamarks Thanks again for reviewing the PR. 
I just wanted to update that if you need to release a new version for confluence project you may go ahead without this PR should the need be. It may take some time to update the docs and write the unit tests. The jupyter at CCAD is also down.
The alternative would be if we could have this PR merged to the new version and I can update the docs and tests later. It might save some time realigning the code with the new version and we have already tested it on the real 500 samples dataset. However, I do see that having the docs and tests updated would be ideal. 
Thanks.

Reply:
@jaamarks Could you please try after this addea7083a996997304920a6c3dabb3c5d7326a8? I have exlcuded using the score field using this incompatibility issue in different ways of preparing bcf. I have tested on my end and it works fine. Let me know if you run into issue. Thanks.

Reply:
@jaamarks 

**For unit testing**

1. I have added three new unit testing scripts scripts in 077f3b1ea5db611e76d87df0be209f8725a939d2 which should test: `src/cgr_gwas_qc/workflow/scripts/vcf2abf.py` , `src/cgr_gwas_qc/workflow/scripts/median_intensity_from_vcf.py ` and `src/cgr_gwas_qc/workflow/scripts/vcf2adpc.py` .
2. I don't see any exisitng testing scripts that checks for any of the `grouped*` script so I have left `src/cgr_gwas_qc/workflow/scripts/grouped_intensity_from_vcf.py` and `src/cgr_gwas_qc/workflow/scripts/grouped_contamination.py`. I believe the rationale might be that these are essentially wrappers to execute main sripts in the parallel mode. I do see that those may be indirectly tested through the subworkflows.
3.  Much of the subworkflows is based on the `RealData`. I have synced the RealData from CCAD2. However, currently even with the default branch the pytest fails on my end. First, I note that  `_reference_fasta ` should be set in RealData since the gtc2vcf workflow is created. Second, I ran into another issue that conda envinorments seem inccorectly setup in this context. I am happy to work through additional subworkflow additions/modifications needed relevent to this PR after the default is fixed.
4. I have updated the FakeData repository with files needed to test bcf entry in a separate PR for the test repo here: https://github.com/NCI-CGR/GwasQcPipeline-test-data/pull/1. Please take a look and merge this first so you may be able to take a look at the new unit testing scripts added in this PR.

Please take a look and let me know if you have any questions. Thank you.

Reply:
@jaamarks 

I have updated the documentation to reflect the BCF entry point:

1. 1afc4e79ebd73e90d01a10ec949800853d9c7783 Indicates the bcf entry point in the user config 
2. cbd7c898fe16a542d61261d8890a359fa8b1ef7a Adds information about bcf entry point on the entry point page. I have added as point in the description since the entry point processing is essentially that it will convert to plink bed (similar to aggregated BED input)
3. 20b6d3074fe7a5fcddb4e1ac46e8ce2c93d2569b Updates the contamination subworkflow documentation to show now it is an isolated subworkflow and supports both GTC and BCF input.
4. b1f1c90ecf8a8c4f7cdaa75cad5594e68b3567c0 and 7236b041d872962592a93f78fe2ffdb536ed155f  Adds the information for intensity subworkflow and show processing for both GTC and BCF inputs.

Please let me know if there is anything else that would suggest to update/modify regarding this PR. Thank you.